fix: fix game screen goes black bug

sid2002CN opened the pull request:
<!-- Thanks for submitting a pull request for Terasology! :-)
Please fill in some details about the PR, below.
If the PR contains source code please make sure to run Checkstyle on it first.
If you add unit tests we'll love you forever! 

You might also want to read "How to Work on a PR Efficiently":
https://github.com/MovingBlocks/Terasology/wiki/How-to-Work-on-a-PR-Efficiently
-->

### Contains

Fixes #4929

### How to test

Play the game and then pause the game.
Minimize the game window.
Bring back the game window and resume the game.

Cervator commented on engine/src/main/java/org/terasology/engine/rendering/opengl/FBO.java:
```diff
@@ -728,6 +728,9 @@ public int height() {
          * @param height An integer representing the new height.
          */
         public void setDimensions(int width, int height) {
+            if(width == 0 && height ==0){
```
Minor code quality fixes here, explained by the robot :-)

Should this just check for an _or_ rather than _and_ ? In the case we're dealing with _both_ dimensions are zero. But it seems like _either_ being zero would be a case where you'd disregard the instruction and return early.

Additionally, maybe it would be good if we did a `logger.warn(....)` to highlight that the set was called with invalid dimensions?

Cervator commented on engine/src/main/java/org/terasology/engine/rendering/opengl/FBO.java:
```diff
@@ -737,6 +740,9 @@ public void setDimensions(int width, int height) {
          * @param other A Dimension to use the width and height from.
          */
         public void setDimensions(Dimensions other) {
+            if(width == 0 && height ==0){
```
In addition to my other review content in this case shouldn't we be checking `other.width` and same for height?

Cervator reviewed (changes requested):
Can confirm this fixes the issue! Mostly, anyway :-)

I replicated the issue before merging, and confirmed the same steps after would no longer encounter the problem - _after_ I dismissed the in-world menu and moved ever so slightly to cause the dimensions to refresh and redraw the screen. However, right as you bring up the game from being minimized the background is still black behind the menu.

So this could _probably_ be fixed more thoroughly, but that's a triviality and this fixes the important bits! Maybe we should leave the issue open as a longer priority to see about finding an "earlier" fix that doesn't cause the blackness to occur in the first place. But I have no idea where to find that, and it isn't very important after this fix :-)

Nice work @sid2002CN ! I have left a couple review comments with trivial adjustments needed. A couple things you could do in the future:

* Make sure you look for warnings in your IDE about style warnings and such, if you're introducing new ones (just missing some spaces in this case). You can run a Checkstyle plugin within IntelliJ for more details
* Consider learning about and using "topic branches" rather than your `develop` branch, for :reasons: that a quick Google search will highlight along with how to do it - don't worry about that for this PR though, just made the tweaks in a second commit in your `develop` branch and we'll merge them!

Thanks!

pollend commented on engine/src/main/java/org/terasology/engine/rendering/opengl/FBO.java:
```diff
@@ -728,6 +728,9 @@ public int height() {
          * @param height An integer representing the new height.
          */
         public void setDimensions(int width, int height) {
```
that's just obscuring the problem by having the method silently fail.

pollend replied:
the precondition should hit and give you a backtrace and give you information where this is coming from. I don't know off the top of my head why its getting set to 0,0

sid2002CN commented:
@Cervator Thanks for your suggestions! This is my first attempt to contribute to an open-source project. I am happy that I have done this in the right way(almost).
As for >finding an "earlier" fix that doesn't cause the blackness to occur in the first place, I kind of get lost on trying to find where to look for. Does anyone know which class actually set the width and height of the FBO?

sid2002CN commented:
Thanks to @Cervator and @pollend's suggestion I found out what actually goes wrong. The width and height of displayDevice are not updated properly as the screen goes back to normal after being minimized. The new solution updates the displayDevice before updating the full-scale if width or height is 0.

Cervator commented:
@sid2002CN very nice detective work! However, oddly enough, I still get the black screen after restoring the window 🤔 Not sure if that is a problem on my end, it would be good with another opinion or two.

Not sure what's happening there, I tried a few different ways to lose focus / minimize and they all result in the same thing. Black behind the pause menu until I dismiss the menu and do anything to trigger a refresh of the world view.

Your fix sounds reasonable, but maybe there are more places to capture? One interesting thing I noticed is that with the game minimized the Windows 10 task bar preview _does_ show the small version of the game window with valid background behind the pause menu. It just doesn't work as I restore the window size - maybe the act of restoring the window is still capturing and storing a 0,0 somewhere?

I don't really know what `displayDevice.update();` does other than what I can assume. Even before testing I wondered if it might still make sense to have the precondition checks _after_ the attempt to run the update, in case there are cases we don't catch.

sid2002CN commented:
@Cervator Thanks for your feedback. I finally got what actually goes wrong. The updateViewport capturing and storing 0,0 during the screen is minimized. The simple solution will be to block the change when the screen is minimized. After spending some time learning how to use GLFW, I come up with this solution.

Cervator commented:
Amazingly well done! I love that you're willing to go this deep into research land to find the proper solution :-)

And I can confirm that it works this time! No more blackness behind the pause menu!

I did however still find an edge case that provokes the original crash though, so there's one more bit of detective work for you 😁 

* Run game, enter world in CoreSampleGameplay
* _Without_ opening the pause menu screen hit the Windows button
* Click a different window to move the focus
* Without otherwise activating the game window click the minimize button on its menu bar
* OG crash!

I dunno if that is capturable similarly or justifies putting the zero check back, now perhaps without the blackness (since it did prevent the crash itself - maybe now it would both not crash _and_ bypass the blackness?). I might suggest having the check there (with preconditions maybe, if that can prevent the crash gentle and be informative) still, even if we resolve the issue. A simple comment explaining briefly why we do the thing would also be helpful

So close now! Thank you very much for the effort here :-)

sid2002CN commented:
@Cervator Never thought of testing it in that way. 😁
It is now fixed with the new commit. It turns out that the game still tries to render the screen after we minimize the game. 
I am not sure how to write in-class comments about this properly. Is there a formal format or I can add comments which simply describe what I have changed?

Cervator commented:
Great to hear about the final fix 👍 

Comment would just be a one liner `// ...` for probably the first change since it includes a property change event and some logic - simply suggest a reason for _why_ we might do / need that., so that it is clear to a reader.

The other change I think documents itself by simply using the well-named boolean. You could read that code without comments as "Ah so if the thing isn't minimized then do this other thing, seems sensible"

sid2002CN commented:
@Cervator Just finished the comment change.

Cervator commented:
Well done - I can't find a way to crash it or see black in the background now. And the comment is on point 👍 

Apologies for taking a little while to get back to this, but the weekend ended 😁 

Approving and merging! Woot!
